linera-storage-service/src/client.rs

```diff
@@ -147,70 +179,82 @@ impl ReadableKeyValueStore<ServiceContextError> for ServiceStoreClient {
         let _guard = self.acquire().await;
         let response = client.process_find_key_values_by_prefix(request).await?;
         let response = response.into_inner();
-        let ReplyFindKeyValuesByPrefix { key_values } = response;
-        let key_values = key_values
-            .into_iter()
-            .map(|x| (x.key, x.value))
-            .collect::<Vec<_>>();
-        Ok(key_values)
+        let ReplyFindKeyValuesByPrefix {
+            key_values,
+            recover_key,
+            n_block,
+        } = response;
+        if n_block == 0 {
+            let key_values = key_values
+                .into_iter()
+                .map(|x| (x.key, x.value))
+                .collect::<Vec<_>>();
+            Ok(key_values)
+        } else {
+            Self::read_entries(client, recover_key, n_block).await
+        }
     }
 }
 
 #[async_trait]
 impl WritableKeyValueStore<ServiceContextError> for ServiceStoreClient {
     const MAX_VALUE_SIZE: usize = usize::MAX;
 
-    async fn write_batch(&self, batch: Batch, base_key: &[u8]) -> Result<(), ServiceContextError> {
+    async fn write_batch(&self, batch: Batch, _base_key: &[u8]) -> Result<(), ServiceContextError> {
         use crate::client::Operation;
         use linera_views::batch::WriteOperation;
         let mut statements = Vec::new();
+        let mut block_size = 0;
         for operation in batch.operations {
-            let operation = match operation {
-                WriteOperation::Delete { key } => {
-                    let mut full_key = self.namespace.clone();
-                    full_key.extend(key);
-                    Operation::Delete(full_key)
-                }
-                WriteOperation::Put { key, value } => {
+            let operation_size = match &operation {
+                WriteOperation::Delete { key } => key.len(),
+                WriteOperation::Put { key, value } => key.len() + value.len(),
+                WriteOperation::DeletePrefix { key_prefix } => key_prefix.len(),
+            };
+            if operation_size + block_size < MAX_GRPC_REQUEST_SIZE {
+                let statement = self.get_statement(operation);
+                statements.push(statement);
+                block_size += operation_size;
+            } else {
+                self.submit_statements(mem::take(&mut statements)).await?;
+                block_size = 0;
+                if operation_size > MAX_GRPC_REQUEST_SIZE {
+                    // One single operation is especially big. So split it in blocks.
+                    let WriteOperation::Put { key, value } = operation else {
+                        // Only the put can go over the limit
+                        unreachable!();
```

Comment:
Does the `Batch` type itself guard against huge keys?

Reply:
No, I added the relevant checks with `ensure`.